### PR TITLE
Support presets for buildspec

### PR DESCRIPTION
**Pull request by krzyzacy**
ref https://github.com/kubernetes/test-infra/pull/11045

/assign @fejta

### CODE REVIEW

**Comment by fejta** [comment]
Alternatively we could remove them from pods! :Impgrin:

**Comment by krzyzacy** [comment]
like what? 😏

**Comment by krzyzacy** [comment]
How kompose will help here?

**Comment by fejta** [comment]
If we want to munge the config, we should be doing this with something like kompose before we upload the configmap into cluster.

We need to stop munging yaml at runtime, whether it be by collating multiple files together or injecting env/volumes.

Presets are bad, bespoke design that no one understands aside from a handful of people, most of whom created the feature.

**Comment by krzyzacy** [comment]
> If we want to munge the config, we should be doing this with something like kompose before we upload the configmap into cluster.

We are doing a lot more things between the raw config map and the return value from config.Config(), like `decorate: true`, and also setting a bunch of default fields, IMO they all have the common goal of reducing boilerplate in our configs.

I'm ok with have a separate controller/job munges the raw config, and spill out some processed config data, and let all other controller consumes the processed config.

> Presets are bad, bespoke design that no one understands aside from a handful of people, most of whom created the feature.

K8s has similar feature so I don't think it's hard to understand, and it's not doing else other than injecting envs/volumes and it also cuts our config size down by half which I think is still a good feature to have. But also feel free to propose alternatives, like how kompose would fit into here.

**Comment by stevekuznetsov** [comment]
I similarly agree with @fejta that the templates we have today are an anti-pattern. For OpenShift, we wrote a Golang generator for job configuration that imports the `test-infra` types and allows our devs to worry about the small subset of fields that actually matter. I don't really think that the size of the Prow config on disk matters and I also don't like us owning a preprocessor for prow config.

`decorate: true` is totally not a fair comparison ... it's _one_ boolean field that hides the complexity from the user.

**Comment by fejta** [comment]
I'm also not a fan of defaulting fields, rather than leaving them nil until we create the job (and complain about it fairly regularly). However this see as substantially less offensive than creating our own runtime templating logic inside of prow.

I do not think decorating is the same thing, as this is more complicated than templating.

**Comment by krzyzacy** [comment]
@stevekuznetsov wonder how openshift handles boilerplate for volumes.

We kinda have a barebone job generator like https://github.com/kubernetes/test-infra/blob/master/experiment/generate_tests.py, and I was chatting with @yguo0905 about make that more generic (not sure if that can replace the preset in the same time, most of them are features in our runner).

Also my goal when integrate with preset was to keep our config small, so if you have alternative ideas which can replace preset and don't explode our configs, feel free to propose, open to discussions :-)

**Comment by fejta** [comment]
If we need we should be using something like jsonnet/ksonnet/bazel/kompose/etc, not our own bespoke thing.

**Comment by stevekuznetsov** [comment]
Why does it matter how large the config is?

**Comment by fejta** [comment]
By large meaning we minimize the amount of repetitive boilerplate job authors need to write

**Comment by krzyzacy** [comment]
maybe give me a link to your goodies @stevekuznetsov ?

**Comment by stevekuznetsov** [comment]
It's [here](https://github.com/openshift/ci-operator-prowgen/blob/master/cmd/ci-operator-prowgen/main.go#L83) and very specific to us, but for instance we take a config like 

```yaml
- artifact_dir: /tmp/artifacts
  as: unit
  commands: ARTIFACT_DIR=/tmp/artifacts make test
  container:
    from: src
```

and turn it into a Presubmit like:

```yaml
  - agent: kubernetes
    always_run: true
    branches:
    - master
    context: ci/prow/unit
    decorate: true
    decoration_config:
      skip_cloning: true
    name: pull-ci-openshift-ci-operator-master-unit
    rerun_command: /test unit
    spec:
      containers:
      - args:
        - --artifact-dir=$(ARTIFACTS)
        - --give-pr-author-access-to-namespace=true
        - --target=unit
        command:
        - ci-operator
        env:
        - name: CONFIG_SPEC
          valueFrom:
            configMapKeyRef:
              key: openshift-ci-operator-master.yaml
              name: ci-operator-configs
        image: ci-operator:latest
        imagePullPolicy: Always
        name: ""
        resources:
          limits:
            cpu: 500m
          requests:
            cpu: 10m
      serviceAccountName: ci-operator
    trigger: ((?m)^/test( all| unit),?(\s+|$))
```

**Comment by krzyzacy** [comment]
So your dev will run the generator?

**Comment by krzyzacy** [comment]
(In the meantime, @fejta mind re-stamp? We can have more design discussions on how the config should look like)

File: prow/config/jobs_test.go [hunk 1]
```diff
@@ -654,7 +675,7 @@ func TestMergePreset(t *testing.T) {
 		},
 	}
 	for _, tc := range tcs {
```
Comment (fejta) [inline prow/config/jobs_test.go]: We need to make sure Steve stays warm:

```golang
for _, tc := range tcs {
  t.Run(tc.name, func(t *testing.T) {
  }
}
```

Then you can remove all the `for test \"%s\"` stuff as this happens automatically

Comment (krzyzacy) [reply prow/config/jobs_test.go]: :+1: done

File: prow/config/config.go [hunk 2]
```diff
@@ -37,6 +37,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/validation"
 	"sigs.k8s.io/yaml"
 
+	buildv1alpha1 "github.com/knative/build/pkg/apis/build/v1alpha1"
```
Comment (fejta) [inline prow/config/config.go]: nit: buildapi is apparently the new way

File: prow/config/config.go [hunk 3]
```diff
@@ -37,6 +37,7 @@ import (
...
 	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
 	prowjobv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
```
Comment (fejta) [inline prow/config/config.go]: why do we import twice?

Comment (krzyzacy) [reply prow/config/config.go]: lol because rebase is dumb? /shrug will fix

Comment (krzyzacy) [reply prow/config/config.go]: lol, actually I blame @stevekuznetsov

**Comment by BenTheElder** [comment]
s/kompose/kustomize, kompose is for docker compose -> kubernetes. we shouldn't intentionally go to the compose format only to transliterate it back.

**Comment by fejta** [comment]
Sure, kompose sounds like the wrong tool in that list (how about we put in skycfg instead?).

My point is I want us to identify, use and recommend one of the many templating technologies in the broader community rather than to create our own.

(Just like how we're replacing run-after-success with knative-pipeline)

File: prow/config/config.go (hunk 3 above)
Comment (stevekuznetsov) [reply prow/config/config.go]: blame tkaen

**Comment by BenTheElder** [comment]
> My point is I want us to identify, use and recommend one of the many templating technologies in the broader community rather than to create our own.

That SGTM but I'm going to point out that OpenShift's generator is the opposite of this and _more_ bespoke than the presets though, FWIW. (not that this is bad! it's just contradictory...)

yaml patches / kustomize are pretty easy to reason about, but we'd need the config to have actual CRD specs, type meta, etc. for this to work properly.

Not sure what bazel templating you're referring to? I don't think ksonnet / jsonnet make sense for this either. Reviewing someone's new prow job as a script that emits it sounds extremely unfun.

**Comment by krzyzacy** [comment]
/hold cancel

https://github.com/kubernetes/test-infra/issues/3338 is now the oldest test-infra issue... let's have follow up discussions

**Pull request merged**